### PR TITLE
runtime: Move String implementation stubs that want need the auto-rel…

### PR DESCRIPTION
**Pull request by aschwaighofer**
…eased return value optimization to an ARC compiled file

String's hashValue function is implemented in terms of Foundation's hash
function in a runtime function on darwin platforms. For non-ASCII strings we
will call str.decomposedStringWithCanonicalMapping inside this runtime function
which will allocate a new NSString and return the result in the current
autorelease pool. We implemented this function in a file compiled without ARC.
This meant that we would leak said NSString into the current active autorelease
pool.
This patch moves the implementation to a file compiled with ARC. ARC will insert
objc_retainAutoreleasedReturnValue call and on platforms that require it an
marker for the hand-off of the autoreleased return value optimization.

SR-4889
rdar://32199117

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: stdlib/public/stubs/SwiftNativeNSXXXBaseARC.m [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#import <Foundation/Foundation.h>
+#import <CoreFoundation/CoreFoundation.h>
+#include <objc/NSObject.h>
+#include <objc/runtime.h>
+#include <objc/objc.h>
+
+/// The following two routines need to be implemented in ARC because
+/// decomposedStringWithCanonicalMapping returns its result autoreleased. And we
+/// want ARC to insert 'objc_retainAutoreleasedReturnValue' and the necessary
+/// markers for the hand-off.
```
Comment (jrose-apple) [inline stdlib/public/stubs/SwiftNativeNSXXXBaseARC.m]: Nitpick: this shouldn't be a doc comment.

File: test/stdlib/StringMemoryTest.swift [hunk 2]
```diff
@@ -0,0 +1,66 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-build-swift -O %s -o %t/StringMemoryTest
+// RUN: %target-run %t/StringMemoryTest | %FileCheck %s
+
+// REQUIRES: optimized_stdlib
+// REQUIRES: executable_test
+// REQUIRES: objc_interop
+
+// The autorelease return optimizaion seems to be failing in the simulator.
+// XFAIL: CPU=i386
```
Comment (jrose-apple) [inline test/stdlib/StringMemoryTest.swift]: I think we just don't do it there. @rjmccall?

File: test/stdlib/runtime_autorelease_optimization.txt [hunk 3]
```diff
@@ -0,0 +1,101 @@
+// UNSUPPORTED: linux
```
Comment (jrose-apple) [inline test/stdlib/runtime_autorelease_optimization.txt]: Better to make this `REQUIRES: objc_interop`. This won't work on Windows or Android either, after all.

File: test/stdlib/StringMemoryTest.swift (hunk 2 above)
Comment (aschwaighofer) [reply test/stdlib/StringMemoryTest.swift]: Yes, seems to be the case. I tried with a pure objective-c app and we also leak there on an i386 simulator rdar://34863458

File: test/stdlib/runtime_autorelease_optimization.txt (hunk 3 above)
Comment (aschwaighofer) [reply test/stdlib/runtime_autorelease_optimization.txt]: Good point

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by gottesmm** [comment]
@aschwaighofer Please be careful about using the word leak and autorelease pool together. Hitting the autorelease pool is not a "leak". I would say "unnecessary memory growth" or something like that. Leak is misleading.

Or am I understanding the pull request, wrong?

**Comment by aschwaighofer** [comment]
@swift-ci Please clean test linux

**Comment by aschwaighofer** [comment]
@swift-ci Please clean test

**Pull request merged**